src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/ReadyToRunCompilationModuleGroupBase.cs

```diff
@@ -0,0 +1,293 @@
+                            }
+                        }
+                        catch (TypeSystemException) { }
+                    }
+                }
+            }
+
+            return _typeRefsInCompilationModuleSet.TryGetValue(type, out token);
+        }
+
+        private bool ComputeTypeVersionsWithCode(TypeDesc type)
+        {
+            if (type == type.Context.CanonType)
+                return true;
+
+            if (type is MetadataType mdType)
+            {
+                if (!_versionBubbleModuleSet.Contains(mdType.Module))
+                    return false;
+            }
+
+            if (type == type.GetTypeDefinition())
+                return true;
+
+            return ComputeInstantiationVersionsWithCode(type.Instantiation, type);
+        }
+
+        private bool ComputeInstantiationVersionsWithCode(Instantiation inst, TypeSystemEntity entityWithInstantiation)
+        {
+            for (int iInstantiation = 0; iInstantiation < inst.Length; iInstantiation++)
+            {
+                TypeDesc instType = inst[iInstantiation];
+
+                if (!ComputeInstantiationTypeVersionsWithCode(instType))
+                {
+                    if (instType.IsPrimitive)
+                    {
+                        // Primitive type instantiations are only instantiated in the module of the generic defining type
+                        // if the generic does not apply interface constraints to that type parameter, or if System.Private.CoreLib is part of the version bubble
+
+                        Instantiation entityDefinitionInstantiation;
+                        if (entityWithInstantiation is TypeDesc type)
+                        {
+                            entityDefinitionInstantiation = type.GetTypeDefinition().Instantiation;
+                        }
+                        else
+                        {
+                            entityDefinitionInstantiation = ((MethodDesc)entityWithInstantiation).GetTypicalMethodDefinition().Instantiation;
+                        }
+
+                        GenericParameterDesc genericParam = (GenericParameterDesc)entityDefinitionInstantiation[iInstantiation];
+                        if (genericParam.Constraints.HasFlag(GenericConstraints.ReferenceTypeConstraint))
+                            return false;
+
+                        // This checks to see if the type constraints list is empty
+                        if (genericParam.TypeConstraints.GetEnumerator().MoveNext())
+                            return false;
+                    }
+                    else
+                    {
+                        // Non-primitive which doesn't version with type implies instantiation doesn't version with type
+                        return false;
+                    }
+                }
+            }
+            return true;
+
+            bool ComputeInstantiationTypeVersionsWithCode(TypeDesc type)
```

Comment:
Nit: I tend to prefix local methods with `static` to avoid any perf gotchas caused by the C# compiler allocating a closure and just don't use the language feature if `static` doesn't cut it.